{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Eflux2 applied to *P. putida*\n",
    "This notebook is based on this [*A. niger* example](https://github.com/AgileBioFoundry/AspergillusQ4Milestone/blob/main/notebooks/Eflux4A.niger.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import cobra\n",
    "import pandas as pd\n",
    "# pd.set_option('display.max_rows', None)\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from optlang import gurobi_interface\n",
    "from optlang.symbolics import add\n",
    "\n",
    "import cplex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('../scripts/')\n",
    "from eflux2 import EFlux2\n",
    "os.chdir('../notebooks')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`../parsed-data/core80_reactions-genes.csv` lists the association between reactions and genes.\n",
    "\n",
    "Make sure that the model you import has the relevant genes associated with each reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2024-10-23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iJN1463</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>2341f2a9f40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>671</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>637</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>596</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*RgPAL - 1.0*RgPAL_reverse_f41ed</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, periplasm, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iJN1463 at 0x2341f2a9f40>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = cobra.io.load_json_model('../models/iJN1463_JS.json')\n",
    "model.solver = 'gurobi'\n",
    "N = cobra.util.create_stoichiometric_matrix(model)\n",
    "model"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import transcriptomics data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>228575.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>547.688969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8478.086255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>39.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>104.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>282.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>776103.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Value\n",
       "count  228575.000000\n",
       "mean      547.688969\n",
       "std      8478.086255\n",
       "min         0.000000\n",
       "25%        39.000000\n",
       "50%       104.000000\n",
       "75%       282.000000\n",
       "max    776103.000000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intensities = pd.read_csv('../data/putida_RNAseq_data.csv').iloc[: , 1:]\n",
    "intensities.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the range of the values are quite large, we perform additive smoothing (i.e. Laplace smoothing)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Value</th>\n",
       "      <th>logValues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>228575.000000</td>\n",
       "      <td>228575.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>547.688969</td>\n",
       "      <td>4.644346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8478.086255</td>\n",
       "      <td>1.583710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.095310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>39.000000</td>\n",
       "      <td>3.691376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>104.000000</td>\n",
       "      <td>4.654912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>282.000000</td>\n",
       "      <td>5.645800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>776103.000000</td>\n",
       "      <td>13.562042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Value      logValues\n",
       "count  228575.000000  228575.000000\n",
       "mean      547.688969       4.644346\n",
       "std      8478.086255       1.583710\n",
       "min         0.000000       0.095310\n",
       "25%        39.000000       3.691376\n",
       "50%       104.000000       4.654912\n",
       "75%       282.000000       5.645800\n",
       "max    776103.000000      13.562042"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADD = 1.1\n",
    "intensities['logValues'] = np.log(intensities['Value'].astype('float') + ADD)\n",
    "intensities.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AroL',\n",
       " 'MCP-SoxS',\n",
       " 'gentamycin-3-acetyltransferase',\n",
       " 'AroG-fbr',\n",
       " 'mob',\n",
       " 'dCas9',\n",
       " 'pBBR1 rep']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[ii for ii in set([i for i in intensities['genes'] if 'PP' not in i]) if ii not in model.genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attributing the nonCRISPR genes to their respective reactions\n",
    "model.reactions.DDPA.gene_reaction_rule = '( PP_1866 and PP_2324 and PP_3080 and AroG-fbr )'\n",
    "model.reactions.SHKK.gene_reaction_rule = '( PP_5079 and AroL )'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strains</th>\n",
       "      <th>4scRNA control</th>\n",
       "      <th>CRISPRa-PP_1075</th>\n",
       "      <th>CRISPRa-PP_1830</th>\n",
       "      <th>CRISPRa-PP_2168</th>\n",
       "      <th>CRISPRa-PP_2329</th>\n",
       "      <th>CRISPRa-PP_4169</th>\n",
       "      <th>CRISPRa-PP_4965</th>\n",
       "      <th>CRISPRa-PP_5079</th>\n",
       "      <th>CRISPRi-PP_0059</th>\n",
       "      <th>CRISPRi-PP_0289</th>\n",
       "      <th>...</th>\n",
       "      <th>CRISPRi-PP_4297</th>\n",
       "      <th>CRISPRi-PP_4678</th>\n",
       "      <th>CRISPRi-PP_4725</th>\n",
       "      <th>CRISPRi-PP_4822</th>\n",
       "      <th>CRISPRi-PP_4823</th>\n",
       "      <th>CRISPRi-PP_5015</th>\n",
       "      <th>CRISPRi-PP_5128</th>\n",
       "      <th>CRISPRi-PP_5155</th>\n",
       "      <th>CRISPRi-PP_5336-5335</th>\n",
       "      <th>pACA production 3 scRNA positive control</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AroG-fbr</th>\n",
       "      <td>10.2037</td>\n",
       "      <td>9.6915</td>\n",
       "      <td>10.2397</td>\n",
       "      <td>10.0788</td>\n",
       "      <td>9.0878</td>\n",
       "      <td>10.1344</td>\n",
       "      <td>10.2923</td>\n",
       "      <td>10.0395</td>\n",
       "      <td>9.7231</td>\n",
       "      <td>10.6007</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0631</td>\n",
       "      <td>9.9279</td>\n",
       "      <td>9.3185</td>\n",
       "      <td>8.5585</td>\n",
       "      <td>8.5647</td>\n",
       "      <td>9.9541</td>\n",
       "      <td>10.0376</td>\n",
       "      <td>8.5274</td>\n",
       "      <td>8.5698</td>\n",
       "      <td>9.7556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AroL</th>\n",
       "      <td>8.4262</td>\n",
       "      <td>7.7459</td>\n",
       "      <td>8.2175</td>\n",
       "      <td>7.8517</td>\n",
       "      <td>7.2815</td>\n",
       "      <td>7.8835</td>\n",
       "      <td>7.9519</td>\n",
       "      <td>8.0017</td>\n",
       "      <td>7.9916</td>\n",
       "      <td>8.3540</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0904</td>\n",
       "      <td>8.0530</td>\n",
       "      <td>7.7134</td>\n",
       "      <td>6.9208</td>\n",
       "      <td>6.9968</td>\n",
       "      <td>8.0784</td>\n",
       "      <td>7.9445</td>\n",
       "      <td>6.8481</td>\n",
       "      <td>7.0131</td>\n",
       "      <td>7.6765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCP-SoxS</th>\n",
       "      <td>6.3904</td>\n",
       "      <td>6.7857</td>\n",
       "      <td>7.4043</td>\n",
       "      <td>7.2690</td>\n",
       "      <td>7.1453</td>\n",
       "      <td>7.0918</td>\n",
       "      <td>7.2464</td>\n",
       "      <td>7.1763</td>\n",
       "      <td>7.5050</td>\n",
       "      <td>8.2207</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1648</td>\n",
       "      <td>6.8201</td>\n",
       "      <td>7.4025</td>\n",
       "      <td>6.8938</td>\n",
       "      <td>6.9461</td>\n",
       "      <td>8.0193</td>\n",
       "      <td>7.1067</td>\n",
       "      <td>6.7276</td>\n",
       "      <td>6.8618</td>\n",
       "      <td>7.1740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0001</th>\n",
       "      <td>5.2988</td>\n",
       "      <td>5.5179</td>\n",
       "      <td>5.2938</td>\n",
       "      <td>5.3476</td>\n",
       "      <td>5.5495</td>\n",
       "      <td>5.3380</td>\n",
       "      <td>5.4557</td>\n",
       "      <td>5.3804</td>\n",
       "      <td>5.7624</td>\n",
       "      <td>6.1593</td>\n",
       "      <td>...</td>\n",
       "      <td>5.2476</td>\n",
       "      <td>5.0882</td>\n",
       "      <td>5.4935</td>\n",
       "      <td>4.8048</td>\n",
       "      <td>5.1990</td>\n",
       "      <td>5.6941</td>\n",
       "      <td>4.8836</td>\n",
       "      <td>5.3088</td>\n",
       "      <td>5.1654</td>\n",
       "      <td>5.5801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0002</th>\n",
       "      <td>5.0046</td>\n",
       "      <td>5.4165</td>\n",
       "      <td>5.1305</td>\n",
       "      <td>5.3986</td>\n",
       "      <td>5.4254</td>\n",
       "      <td>5.0505</td>\n",
       "      <td>5.4254</td>\n",
       "      <td>5.0113</td>\n",
       "      <td>5.6316</td>\n",
       "      <td>5.8233</td>\n",
       "      <td>...</td>\n",
       "      <td>5.4076</td>\n",
       "      <td>5.2155</td>\n",
       "      <td>5.4254</td>\n",
       "      <td>4.8683</td>\n",
       "      <td>5.1066</td>\n",
       "      <td>5.5495</td>\n",
       "      <td>5.1654</td>\n",
       "      <td>4.9352</td>\n",
       "      <td>4.9843</td>\n",
       "      <td>5.5219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0003</th>\n",
       "      <td>4.5860</td>\n",
       "      <td>4.5654</td>\n",
       "      <td>4.8911</td>\n",
       "      <td>4.7104</td>\n",
       "      <td>4.6357</td>\n",
       "      <td>4.1287</td>\n",
       "      <td>4.7194</td>\n",
       "      <td>4.5009</td>\n",
       "      <td>4.8606</td>\n",
       "      <td>4.7283</td>\n",
       "      <td>...</td>\n",
       "      <td>4.4671</td>\n",
       "      <td>4.3451</td>\n",
       "      <td>4.7715</td>\n",
       "      <td>4.0091</td>\n",
       "      <td>4.4555</td>\n",
       "      <td>4.9207</td>\n",
       "      <td>4.4898</td>\n",
       "      <td>4.3320</td>\n",
       "      <td>4.1759</td>\n",
       "      <td>4.7458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0004</th>\n",
       "      <td>6.8629</td>\n",
       "      <td>7.1165</td>\n",
       "      <td>6.9557</td>\n",
       "      <td>6.9432</td>\n",
       "      <td>7.2241</td>\n",
       "      <td>6.9538</td>\n",
       "      <td>7.0140</td>\n",
       "      <td>7.0336</td>\n",
       "      <td>7.2725</td>\n",
       "      <td>7.1960</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1042</td>\n",
       "      <td>6.8003</td>\n",
       "      <td>7.1157</td>\n",
       "      <td>6.6175</td>\n",
       "      <td>6.6872</td>\n",
       "      <td>7.1663</td>\n",
       "      <td>7.0185</td>\n",
       "      <td>6.8470</td>\n",
       "      <td>6.5696</td>\n",
       "      <td>7.2094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0005</th>\n",
       "      <td>6.2617</td>\n",
       "      <td>6.0687</td>\n",
       "      <td>6.0405</td>\n",
       "      <td>5.9842</td>\n",
       "      <td>6.0429</td>\n",
       "      <td>6.0847</td>\n",
       "      <td>6.0938</td>\n",
       "      <td>5.7465</td>\n",
       "      <td>5.8639</td>\n",
       "      <td>5.6702</td>\n",
       "      <td>...</td>\n",
       "      <td>5.6563</td>\n",
       "      <td>5.7841</td>\n",
       "      <td>5.5138</td>\n",
       "      <td>5.7108</td>\n",
       "      <td>5.9138</td>\n",
       "      <td>5.7686</td>\n",
       "      <td>5.7749</td>\n",
       "      <td>5.6702</td>\n",
       "      <td>5.6839</td>\n",
       "      <td>6.1423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0006</th>\n",
       "      <td>6.4601</td>\n",
       "      <td>6.6491</td>\n",
       "      <td>6.4054</td>\n",
       "      <td>6.3046</td>\n",
       "      <td>6.9168</td>\n",
       "      <td>6.6696</td>\n",
       "      <td>6.6162</td>\n",
       "      <td>6.7812</td>\n",
       "      <td>7.0519</td>\n",
       "      <td>7.3032</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7880</td>\n",
       "      <td>6.3718</td>\n",
       "      <td>6.5059</td>\n",
       "      <td>6.3684</td>\n",
       "      <td>6.1740</td>\n",
       "      <td>6.6295</td>\n",
       "      <td>6.3511</td>\n",
       "      <td>6.3046</td>\n",
       "      <td>6.2540</td>\n",
       "      <td>6.6569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0007</th>\n",
       "      <td>1.8083</td>\n",
       "      <td>3.0007</td>\n",
       "      <td>2.0919</td>\n",
       "      <td>1.9601</td>\n",
       "      <td>2.4932</td>\n",
       "      <td>1.1314</td>\n",
       "      <td>2.4932</td>\n",
       "      <td>2.7788</td>\n",
       "      <td>3.0007</td>\n",
       "      <td>2.0919</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2083</td>\n",
       "      <td>2.3125</td>\n",
       "      <td>2.2083</td>\n",
       "      <td>2.0919</td>\n",
       "      <td>2.5726</td>\n",
       "      <td>2.8391</td>\n",
       "      <td>2.5726</td>\n",
       "      <td>2.4932</td>\n",
       "      <td>2.6462</td>\n",
       "      <td>2.2083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "strains   4scRNA control  CRISPRa-PP_1075  CRISPRa-PP_1830  CRISPRa-PP_2168  \\\n",
       "genes                                                                         \n",
       "AroG-fbr         10.2037           9.6915          10.2397          10.0788   \n",
       "AroL              8.4262           7.7459           8.2175           7.8517   \n",
       "MCP-SoxS          6.3904           6.7857           7.4043           7.2690   \n",
       "PP_0001           5.2988           5.5179           5.2938           5.3476   \n",
       "PP_0002           5.0046           5.4165           5.1305           5.3986   \n",
       "PP_0003           4.5860           4.5654           4.8911           4.7104   \n",
       "PP_0004           6.8629           7.1165           6.9557           6.9432   \n",
       "PP_0005           6.2617           6.0687           6.0405           5.9842   \n",
       "PP_0006           6.4601           6.6491           6.4054           6.3046   \n",
       "PP_0007           1.8083           3.0007           2.0919           1.9601   \n",
       "\n",
       "strains   CRISPRa-PP_2329  CRISPRa-PP_4169  CRISPRa-PP_4965  CRISPRa-PP_5079  \\\n",
       "genes                                                                          \n",
       "AroG-fbr           9.0878          10.1344          10.2923          10.0395   \n",
       "AroL               7.2815           7.8835           7.9519           8.0017   \n",
       "MCP-SoxS           7.1453           7.0918           7.2464           7.1763   \n",
       "PP_0001            5.5495           5.3380           5.4557           5.3804   \n",
       "PP_0002            5.4254           5.0505           5.4254           5.0113   \n",
       "PP_0003            4.6357           4.1287           4.7194           4.5009   \n",
       "PP_0004            7.2241           6.9538           7.0140           7.0336   \n",
       "PP_0005            6.0429           6.0847           6.0938           5.7465   \n",
       "PP_0006            6.9168           6.6696           6.6162           6.7812   \n",
       "PP_0007            2.4932           1.1314           2.4932           2.7788   \n",
       "\n",
       "strains   CRISPRi-PP_0059  CRISPRi-PP_0289  ...  CRISPRi-PP_4297  \\\n",
       "genes                                       ...                    \n",
       "AroG-fbr           9.7231          10.6007  ...          10.0631   \n",
       "AroL               7.9916           8.3540  ...           8.0904   \n",
       "MCP-SoxS           7.5050           8.2207  ...           7.1648   \n",
       "PP_0001            5.7624           6.1593  ...           5.2476   \n",
       "PP_0002            5.6316           5.8233  ...           5.4076   \n",
       "PP_0003            4.8606           4.7283  ...           4.4671   \n",
       "PP_0004            7.2725           7.1960  ...           7.1042   \n",
       "PP_0005            5.8639           5.6702  ...           5.6563   \n",
       "PP_0006            7.0519           7.3032  ...           6.7880   \n",
       "PP_0007            3.0007           2.0919  ...           2.2083   \n",
       "\n",
       "strains   CRISPRi-PP_4678  CRISPRi-PP_4725  CRISPRi-PP_4822  CRISPRi-PP_4823  \\\n",
       "genes                                                                          \n",
       "AroG-fbr           9.9279           9.3185           8.5585           8.5647   \n",
       "AroL               8.0530           7.7134           6.9208           6.9968   \n",
       "MCP-SoxS           6.8201           7.4025           6.8938           6.9461   \n",
       "PP_0001            5.0882           5.4935           4.8048           5.1990   \n",
       "PP_0002            5.2155           5.4254           4.8683           5.1066   \n",
       "PP_0003            4.3451           4.7715           4.0091           4.4555   \n",
       "PP_0004            6.8003           7.1157           6.6175           6.6872   \n",
       "PP_0005            5.7841           5.5138           5.7108           5.9138   \n",
       "PP_0006            6.3718           6.5059           6.3684           6.1740   \n",
       "PP_0007            2.3125           2.2083           2.0919           2.5726   \n",
       "\n",
       "strains   CRISPRi-PP_5015  CRISPRi-PP_5128  CRISPRi-PP_5155  \\\n",
       "genes                                                         \n",
       "AroG-fbr           9.9541          10.0376           8.5274   \n",
       "AroL               8.0784           7.9445           6.8481   \n",
       "MCP-SoxS           8.0193           7.1067           6.7276   \n",
       "PP_0001            5.6941           4.8836           5.3088   \n",
       "PP_0002            5.5495           5.1654           4.9352   \n",
       "PP_0003            4.9207           4.4898           4.3320   \n",
       "PP_0004            7.1663           7.0185           6.8470   \n",
       "PP_0005            5.7686           5.7749           5.6702   \n",
       "PP_0006            6.6295           6.3511           6.3046   \n",
       "PP_0007            2.8391           2.5726           2.4932   \n",
       "\n",
       "strains   CRISPRi-PP_5336-5335  pACA production 3 scRNA positive control  \n",
       "genes                                                                     \n",
       "AroG-fbr                8.5698                                    9.7556  \n",
       "AroL                    7.0131                                    7.6765  \n",
       "MCP-SoxS                6.8618                                    7.1740  \n",
       "PP_0001                 5.1654                                    5.5801  \n",
       "PP_0002                 4.9843                                    5.5219  \n",
       "PP_0003                 4.1759                                    4.7458  \n",
       "PP_0004                 6.5696                                    7.2094  \n",
       "PP_0005                 5.6839                                    6.1423  \n",
       "PP_0006                 6.2540                                    6.6569  \n",
       "PP_0007                 2.6462                                    2.2083  \n",
       "\n",
       "[10 rows x 41 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_data = pd.pivot_table(intensities, values='logValues', index='genes', columns='strains').round(4)\n",
    "cleaned_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strains</th>\n",
       "      <th>4scRNA control</th>\n",
       "      <th>CRISPRa-PP_1075</th>\n",
       "      <th>CRISPRa-PP_1830</th>\n",
       "      <th>CRISPRa-PP_2168</th>\n",
       "      <th>CRISPRa-PP_2329</th>\n",
       "      <th>CRISPRa-PP_4169</th>\n",
       "      <th>CRISPRa-PP_4965</th>\n",
       "      <th>CRISPRa-PP_5079</th>\n",
       "      <th>CRISPRi-PP_0059</th>\n",
       "      <th>CRISPRi-PP_0289</th>\n",
       "      <th>...</th>\n",
       "      <th>CRISPRi-PP_4297</th>\n",
       "      <th>CRISPRi-PP_4678</th>\n",
       "      <th>CRISPRi-PP_4725</th>\n",
       "      <th>CRISPRi-PP_4822</th>\n",
       "      <th>CRISPRi-PP_4823</th>\n",
       "      <th>CRISPRi-PP_5015</th>\n",
       "      <th>CRISPRi-PP_5128</th>\n",
       "      <th>CRISPRi-PP_5155</th>\n",
       "      <th>CRISPRi-PP_5336-5335</th>\n",
       "      <th>pACA production 3 scRNA positive control</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AroG-fbr</th>\n",
       "      <td>0.962550</td>\n",
       "      <td>0.914232</td>\n",
       "      <td>0.965946</td>\n",
       "      <td>0.950767</td>\n",
       "      <td>0.857283</td>\n",
       "      <td>0.956012</td>\n",
       "      <td>0.970908</td>\n",
       "      <td>0.947060</td>\n",
       "      <td>0.917213</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.949286</td>\n",
       "      <td>0.936532</td>\n",
       "      <td>0.879046</td>\n",
       "      <td>0.807352</td>\n",
       "      <td>0.807937</td>\n",
       "      <td>0.939004</td>\n",
       "      <td>0.946881</td>\n",
       "      <td>0.804419</td>\n",
       "      <td>0.808418</td>\n",
       "      <td>0.920279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AroL</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.919264</td>\n",
       "      <td>0.975232</td>\n",
       "      <td>0.931820</td>\n",
       "      <td>0.864150</td>\n",
       "      <td>0.935594</td>\n",
       "      <td>0.943711</td>\n",
       "      <td>0.949621</td>\n",
       "      <td>0.948423</td>\n",
       "      <td>0.991431</td>\n",
       "      <td>...</td>\n",
       "      <td>0.960148</td>\n",
       "      <td>0.955710</td>\n",
       "      <td>0.915407</td>\n",
       "      <td>0.821343</td>\n",
       "      <td>0.830362</td>\n",
       "      <td>0.958724</td>\n",
       "      <td>0.942833</td>\n",
       "      <td>0.812715</td>\n",
       "      <td>0.832297</td>\n",
       "      <td>0.911028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCP-SoxS</th>\n",
       "      <td>0.777355</td>\n",
       "      <td>0.825441</td>\n",
       "      <td>0.900690</td>\n",
       "      <td>0.884231</td>\n",
       "      <td>0.869184</td>\n",
       "      <td>0.862676</td>\n",
       "      <td>0.881482</td>\n",
       "      <td>0.872955</td>\n",
       "      <td>0.912939</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.871556</td>\n",
       "      <td>0.829625</td>\n",
       "      <td>0.900471</td>\n",
       "      <td>0.838590</td>\n",
       "      <td>0.844952</td>\n",
       "      <td>0.975501</td>\n",
       "      <td>0.864488</td>\n",
       "      <td>0.818373</td>\n",
       "      <td>0.834698</td>\n",
       "      <td>0.872675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0001</th>\n",
       "      <td>0.860293</td>\n",
       "      <td>0.895865</td>\n",
       "      <td>0.859481</td>\n",
       "      <td>0.868216</td>\n",
       "      <td>0.900995</td>\n",
       "      <td>0.866657</td>\n",
       "      <td>0.885766</td>\n",
       "      <td>0.873541</td>\n",
       "      <td>0.935561</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.851980</td>\n",
       "      <td>0.826100</td>\n",
       "      <td>0.891903</td>\n",
       "      <td>0.780089</td>\n",
       "      <td>0.844089</td>\n",
       "      <td>0.924472</td>\n",
       "      <td>0.792882</td>\n",
       "      <td>0.861916</td>\n",
       "      <td>0.838634</td>\n",
       "      <td>0.905963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0002</th>\n",
       "      <td>0.859410</td>\n",
       "      <td>0.930143</td>\n",
       "      <td>0.881030</td>\n",
       "      <td>0.927069</td>\n",
       "      <td>0.931671</td>\n",
       "      <td>0.867292</td>\n",
       "      <td>0.931671</td>\n",
       "      <td>0.860560</td>\n",
       "      <td>0.967081</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.928614</td>\n",
       "      <td>0.895626</td>\n",
       "      <td>0.931671</td>\n",
       "      <td>0.836004</td>\n",
       "      <td>0.876925</td>\n",
       "      <td>0.952982</td>\n",
       "      <td>0.887023</td>\n",
       "      <td>0.847492</td>\n",
       "      <td>0.855924</td>\n",
       "      <td>0.948242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0003</th>\n",
       "      <td>0.917585</td>\n",
       "      <td>0.913464</td>\n",
       "      <td>0.978631</td>\n",
       "      <td>0.942476</td>\n",
       "      <td>0.927530</td>\n",
       "      <td>0.826087</td>\n",
       "      <td>0.944277</td>\n",
       "      <td>0.900558</td>\n",
       "      <td>0.972528</td>\n",
       "      <td>0.946057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.893795</td>\n",
       "      <td>0.869385</td>\n",
       "      <td>0.954701</td>\n",
       "      <td>0.802157</td>\n",
       "      <td>0.891474</td>\n",
       "      <td>0.984554</td>\n",
       "      <td>0.898337</td>\n",
       "      <td>0.866764</td>\n",
       "      <td>0.835531</td>\n",
       "      <td>0.949559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0004</th>\n",
       "      <td>0.941891</td>\n",
       "      <td>0.976696</td>\n",
       "      <td>0.954627</td>\n",
       "      <td>0.952912</td>\n",
       "      <td>0.991463</td>\n",
       "      <td>0.954366</td>\n",
       "      <td>0.962628</td>\n",
       "      <td>0.965318</td>\n",
       "      <td>0.998106</td>\n",
       "      <td>0.987607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.975008</td>\n",
       "      <td>0.933299</td>\n",
       "      <td>0.976586</td>\n",
       "      <td>0.908211</td>\n",
       "      <td>0.917777</td>\n",
       "      <td>0.983531</td>\n",
       "      <td>0.963246</td>\n",
       "      <td>0.939709</td>\n",
       "      <td>0.901637</td>\n",
       "      <td>0.989446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0005</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.969178</td>\n",
       "      <td>0.964674</td>\n",
       "      <td>0.955683</td>\n",
       "      <td>0.965057</td>\n",
       "      <td>0.971733</td>\n",
       "      <td>0.973186</td>\n",
       "      <td>0.917722</td>\n",
       "      <td>0.936471</td>\n",
       "      <td>0.905537</td>\n",
       "      <td>...</td>\n",
       "      <td>0.903317</td>\n",
       "      <td>0.923727</td>\n",
       "      <td>0.880560</td>\n",
       "      <td>0.912021</td>\n",
       "      <td>0.944440</td>\n",
       "      <td>0.921251</td>\n",
       "      <td>0.922258</td>\n",
       "      <td>0.905537</td>\n",
       "      <td>0.907725</td>\n",
       "      <td>0.980932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0006</th>\n",
       "      <td>0.884557</td>\n",
       "      <td>0.910437</td>\n",
       "      <td>0.877068</td>\n",
       "      <td>0.863265</td>\n",
       "      <td>0.947092</td>\n",
       "      <td>0.913244</td>\n",
       "      <td>0.905932</td>\n",
       "      <td>0.928524</td>\n",
       "      <td>0.965590</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.929456</td>\n",
       "      <td>0.872467</td>\n",
       "      <td>0.890829</td>\n",
       "      <td>0.872001</td>\n",
       "      <td>0.845383</td>\n",
       "      <td>0.907753</td>\n",
       "      <td>0.869632</td>\n",
       "      <td>0.863265</td>\n",
       "      <td>0.856337</td>\n",
       "      <td>0.911505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP_0007</th>\n",
       "      <td>0.584152</td>\n",
       "      <td>0.969344</td>\n",
       "      <td>0.675766</td>\n",
       "      <td>0.633189</td>\n",
       "      <td>0.805401</td>\n",
       "      <td>0.365486</td>\n",
       "      <td>0.805401</td>\n",
       "      <td>0.897661</td>\n",
       "      <td>0.969344</td>\n",
       "      <td>0.675766</td>\n",
       "      <td>...</td>\n",
       "      <td>0.713367</td>\n",
       "      <td>0.747028</td>\n",
       "      <td>0.713367</td>\n",
       "      <td>0.675766</td>\n",
       "      <td>0.831051</td>\n",
       "      <td>0.917140</td>\n",
       "      <td>0.831051</td>\n",
       "      <td>0.805401</td>\n",
       "      <td>0.854826</td>\n",
       "      <td>0.713367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "strains   4scRNA control  CRISPRa-PP_1075  CRISPRa-PP_1830  CRISPRa-PP_2168  \\\n",
       "genes                                                                         \n",
       "AroG-fbr        0.962550         0.914232         0.965946         0.950767   \n",
       "AroL            1.000000         0.919264         0.975232         0.931820   \n",
       "MCP-SoxS        0.777355         0.825441         0.900690         0.884231   \n",
       "PP_0001         0.860293         0.895865         0.859481         0.868216   \n",
       "PP_0002         0.859410         0.930143         0.881030         0.927069   \n",
       "PP_0003         0.917585         0.913464         0.978631         0.942476   \n",
       "PP_0004         0.941891         0.976696         0.954627         0.952912   \n",
       "PP_0005         1.000000         0.969178         0.964674         0.955683   \n",
       "PP_0006         0.884557         0.910437         0.877068         0.863265   \n",
       "PP_0007         0.584152         0.969344         0.675766         0.633189   \n",
       "\n",
       "strains   CRISPRa-PP_2329  CRISPRa-PP_4169  CRISPRa-PP_4965  CRISPRa-PP_5079  \\\n",
       "genes                                                                          \n",
       "AroG-fbr         0.857283         0.956012         0.970908         0.947060   \n",
       "AroL             0.864150         0.935594         0.943711         0.949621   \n",
       "MCP-SoxS         0.869184         0.862676         0.881482         0.872955   \n",
       "PP_0001          0.900995         0.866657         0.885766         0.873541   \n",
       "PP_0002          0.931671         0.867292         0.931671         0.860560   \n",
       "PP_0003          0.927530         0.826087         0.944277         0.900558   \n",
       "PP_0004          0.991463         0.954366         0.962628         0.965318   \n",
       "PP_0005          0.965057         0.971733         0.973186         0.917722   \n",
       "PP_0006          0.947092         0.913244         0.905932         0.928524   \n",
       "PP_0007          0.805401         0.365486         0.805401         0.897661   \n",
       "\n",
       "strains   CRISPRi-PP_0059  CRISPRi-PP_0289  ...  CRISPRi-PP_4297  \\\n",
       "genes                                       ...                    \n",
       "AroG-fbr         0.917213         1.000000  ...         0.949286   \n",
       "AroL             0.948423         0.991431  ...         0.960148   \n",
       "MCP-SoxS         0.912939         1.000000  ...         0.871556   \n",
       "PP_0001          0.935561         1.000000  ...         0.851980   \n",
       "PP_0002          0.967081         1.000000  ...         0.928614   \n",
       "PP_0003          0.972528         0.946057  ...         0.893795   \n",
       "PP_0004          0.998106         0.987607  ...         0.975008   \n",
       "PP_0005          0.936471         0.905537  ...         0.903317   \n",
       "PP_0006          0.965590         1.000000  ...         0.929456   \n",
       "PP_0007          0.969344         0.675766  ...         0.713367   \n",
       "\n",
       "strains   CRISPRi-PP_4678  CRISPRi-PP_4725  CRISPRi-PP_4822  CRISPRi-PP_4823  \\\n",
       "genes                                                                          \n",
       "AroG-fbr         0.936532         0.879046         0.807352         0.807937   \n",
       "AroL             0.955710         0.915407         0.821343         0.830362   \n",
       "MCP-SoxS         0.829625         0.900471         0.838590         0.844952   \n",
       "PP_0001          0.826100         0.891903         0.780089         0.844089   \n",
       "PP_0002          0.895626         0.931671         0.836004         0.876925   \n",
       "PP_0003          0.869385         0.954701         0.802157         0.891474   \n",
       "PP_0004          0.933299         0.976586         0.908211         0.917777   \n",
       "PP_0005          0.923727         0.880560         0.912021         0.944440   \n",
       "PP_0006          0.872467         0.890829         0.872001         0.845383   \n",
       "PP_0007          0.747028         0.713367         0.675766         0.831051   \n",
       "\n",
       "strains   CRISPRi-PP_5015  CRISPRi-PP_5128  CRISPRi-PP_5155  \\\n",
       "genes                                                         \n",
       "AroG-fbr         0.939004         0.946881         0.804419   \n",
       "AroL             0.958724         0.942833         0.812715   \n",
       "MCP-SoxS         0.975501         0.864488         0.818373   \n",
       "PP_0001          0.924472         0.792882         0.861916   \n",
       "PP_0002          0.952982         0.887023         0.847492   \n",
       "PP_0003          0.984554         0.898337         0.866764   \n",
       "PP_0004          0.983531         0.963246         0.939709   \n",
       "PP_0005          0.921251         0.922258         0.905537   \n",
       "PP_0006          0.907753         0.869632         0.863265   \n",
       "PP_0007          0.917140         0.831051         0.805401   \n",
       "\n",
       "strains   CRISPRi-PP_5336-5335  pACA production 3 scRNA positive control  \n",
       "genes                                                                     \n",
       "AroG-fbr              0.808418                                  0.920279  \n",
       "AroL                  0.832297                                  0.911028  \n",
       "MCP-SoxS              0.834698                                  0.872675  \n",
       "PP_0001               0.838634                                  0.905963  \n",
       "PP_0002               0.855924                                  0.948242  \n",
       "PP_0003               0.835531                                  0.949559  \n",
       "PP_0004               0.901637                                  0.989446  \n",
       "PP_0005               0.907725                                  0.980932  \n",
       "PP_0006               0.856337                                  0.911505  \n",
       "PP_0007               0.854826                                  0.713367  \n",
       "\n",
       "[10 rows x 41 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find max of each row\n",
    "cleaned_data.max(axis=1)\n",
    "# normalize each row by dividing by the max possible value\n",
    "norm_data = cleaned_data.divide(cleaned_data.max(axis=1), axis=0)\n",
    "# multiply by glucose intake\n",
    "    # model.metabolites.glc__D_e.summary()\n",
    "    # norm_data = norm_data * 6\n",
    "\n",
    "norm_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genes\n",
       "PP_2842    3.4689\n",
       "PP_2843    1.9601\n",
       "PP_2844    0.0953\n",
       "PP_2845    4.6738\n",
       "PP_2846    2.0919\n",
       "PP_2847    2.4069\n",
       "PP_2848    2.4932\n",
       "PP_2849    3.5863\n",
       "Name: CRISPRi-PP_0289, dtype: float64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for UREA reaction\n",
    "cleaned_data.loc['PP_2842':'PP_2849']['CRISPRi-PP_0289']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use housekeeping genes to determine the variability in measurements\n",
    "source: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6765009/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.03171450622820654"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import statistics\n",
    "\n",
    "hkGenes = ['PP_0013', 'PP_5089','PP_0603', 'PP_4121', 'PP_2082', 'PP_1629', 'PP_0447', 'PP_0387']\n",
    "stdevs = []\n",
    "for i in hkGenes:\n",
    "    stdevs.append(statistics.pstdev(norm_data.loc[i].values))\n",
    "\n",
    "statistics.mean(stdevs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## troubleshooting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## helper functions\n",
    "def create_gprdict(model):   \n",
    "    gpr_dict = dict()\n",
    "    for rxn in model.reactions:\n",
    "        if rxn.gene_reaction_rule:\n",
    "            temp = set()\n",
    "            for x in [x.strip('() ') for x in rxn.gene_reaction_rule.split(' or ')]:\n",
    "                temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "            gpr_dict[rxn.id] = temp\n",
    "    return gpr_dict\n",
    "\n",
    "def transcript_value_for_rxn(model, transcriptomics_df, rxn):\n",
    "    final_transcript_value = 0\n",
    "    gene_ids = []\n",
    "    for parallel_gene in create_gprdict(model)[rxn.id]:\n",
    "        transcript_values = []\n",
    "        for gene in parallel_gene:\n",
    "            if gene in transcriptomics_df.index:\n",
    "                transcript_values.append(transcriptomics_df.loc[gene])\n",
    "                # transcript_values.append(transcriptomics_df.loc[gene].to_numpy()[0])\n",
    "#                 print(transcriptomics_df.loc[gene].to_numpy()[0])\n",
    "            else:\n",
    "                transcript_values.append(np.inf)\n",
    "            min_transcript_val = np.min(transcript_values)\n",
    "        final_transcript_value = final_transcript_value + min_transcript_val\n",
    "#         if final_transcript_value==newinfbound:\n",
    "#             display(rxn.id)\n",
    "#             gene_ids.append(rxn.id)\n",
    "    return final_transcript_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmphe1w4jfb.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "1.5730485607153426\n",
      "FBA status: optimal\n",
      "FBA solution: 1.5730485607153426\n"
     ]
    }
   ],
   "source": [
    "Transcriptomics = norm_data['4scRNA control']\n",
    "model2 = model.copy()\n",
    "model2.tolerance = 1e-9\n",
    "sol = model.optimize()\n",
    "print(sol['RgPAL'])\n",
    "\n",
    "for rxn in model2.reactions:\n",
    "    if rxn.id in ['FBA','PGI', 'RgPAL']:\n",
    "        pass\n",
    "    elif 'EX_' not in str(rxn.id):\n",
    "        if rxn.gene_reaction_rule:\n",
    "            # print(rxn.id)\n",
    "            if rxn.upper_bound > 0.0:\n",
    "                rxn.upper_bound = transcript_value_for_rxn(model, Transcriptomics, rxn) * rxn.upper_bound\n",
    "            else:\n",
    "                pass\n",
    "            if rxn.lower_bound < 0.0:\n",
    "                rxn.lower_bound = transcript_value_for_rxn(model, Transcriptomics, rxn) * rxn.lower_bound\n",
    "            else:\n",
    "                pass\n",
    "                \n",
    "        else:\n",
    "            \"\"\"When there is no GPR, the arbitrary bounds are removed. \n",
    "            Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\"\"\"\n",
    "            if rxn.lower_bound <= -np.log(1000):\n",
    "                rxn.lower_bound = -np.Inf\n",
    "            if rxn.upper_bound >= np.log(1000):\n",
    "                rxn.upper_bound = np.Inf \n",
    "\n",
    "# solve FBA problem with transcriptomic bounds\n",
    "fba_solution = model2.optimize()\n",
    "print('FBA status:', fba_solution.status)\n",
    "print('FBA solution:', fba_solution.objective_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIOMASS_KT2440_WT3 lowerbound\n",
      "EX_glc__D_e lowerbound\n",
      "FACOAL161t2pp lowerbound\n",
      "FBA upperbound\n",
      "G6PDH2r lowerbound\n",
      "HACD21i lowerbound\n",
      "KAT21 lowerbound\n",
      "LPLIPAL1G161pp lowerbound\n",
      "LPLIPAL1G181d11pp lowerbound\n",
      "MLDCP2Bpp lowerbound\n",
      "MLTGY2pp lowerbound\n",
      "PLIPA2G180pp lowerbound\n",
      "PLIPA2G181d11pp lowerbound\n",
      "AGM4PCPpp lowerbound\n",
      "ATPM upperbound\n",
      "ATPM lowerbound\n",
      "CHRPL lowerbound\n",
      "EX_4abz_e lowerbound\n",
      "EX_4aPhe_e lowerbound\n",
      "PAL_F lowerbound\n",
      "PAL_Y lowerbound\n"
     ]
    }
   ],
   "source": [
    "fba_solution\n",
    "# for each reaction in the model\n",
    "for i in model2.reactions:\n",
    "# if the fba solution is equal to the upper bound, then print reaction id and \"upperbound\"\n",
    "    if fba_solution[i.id]==i.upper_bound:\n",
    "        print(i.id, \"upperbound\")\n",
    "# if the fba solution is equal to the lower bound, then print reaction id and \"lowerbound\"\n",
    "    if fba_solution[i.id]==i.lower_bound:\n",
    "        print(i.id, \"lowerbound\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The biomass, FBA, G6PDH2r, PAL_F, and PAL_Y reactions are constrained so that they do not fall to zero. \n",
    "ATPM is constrained to a single literature value. \n",
    "Thus, the only reaction that is worth constraining with our transcriptomic data is RgPAL. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Eflux2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runEflux2(Transcriptomics, model):\n",
    "    model2 = model.copy()\n",
    "    model2.tolerance = 1e-9\n",
    "    sol = model.optimize()\n",
    "    \n",
    "    for rxn in model2.reactions:\n",
    "        if rxn.id in ['FBA','PGI', 'RgPAL']:\n",
    "            pass\n",
    "        elif 'EX_' not in str(rxn.id):\n",
    "            if rxn.gene_reaction_rule:\n",
    "                # print(rxn.id)\n",
    "                if rxn.upper_bound > 0.0:\n",
    "                    rxn.upper_bound = transcript_value_for_rxn(model, Transcriptomics, rxn) * rxn.upper_bound\n",
    "                else:\n",
    "                    pass\n",
    "                if rxn.lower_bound < 0.0:\n",
    "                    rxn.lower_bound = transcript_value_for_rxn(model, Transcriptomics, rxn) * rxn.lower_bound\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "            else:\n",
    "                \"\"\"When there is no GPR, the arbitrary bounds are removed. \n",
    "                Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\"\"\"\n",
    "                if rxn.lower_bound <= -np.log(1000):\n",
    "                    rxn.lower_bound = -np.Inf\n",
    "                if rxn.upper_bound >= np.log(1000):\n",
    "                    rxn.upper_bound = np.Inf \n",
    "    \n",
    "    model2.reactions.RgPAL.upper_bound = transcript_value_for_rxn(model, Transcriptomics, model.reactions.RgPAL) * sol['RgPAL']\n",
    "    # solve FBA problem with transcriptomic bounds\n",
    "    fba_solution = model2.optimize()\n",
    "    # print('FBA status:', fba_solution.status)\n",
    "    print('FBA solution:', fba_solution.objective_value)\n",
    "    \n",
    "    # constrain the biomass to the optimal value\n",
    "    for r in model2.reactions:\n",
    "        if r.objective_coefficient:\n",
    "            try:\n",
    "                r.lower_bound = fba_solution.objective_value\n",
    "                r.upper_bound = fba_solution.objective_value\n",
    "            except:\n",
    "                r.upper_bound = fba_solution.objective_value\n",
    "                r.lower_bound = fba_solution.objective_value\n",
    "\n",
    "    # Inspect media\n",
    "#     display(eflux2_model.medium)\n",
    "    \n",
    "    # Minimize the sum of squared flux values\n",
    "    \"\"\"Note: Because of quadratic objective still have to use cplex objective formulation.\n",
    "    Optlang does not support quadratic type of constraints and objectives yet.\"\"\"\n",
    "    model2.objective = model2.problem.Objective(add([rxn.flux_expression**2 for rxn in model2.reactions]), direction='min')\n",
    "    \n",
    "    # solve the minimization of squared fluxes problem\n",
    "    EFlux2_solution = model2.optimize()\n",
    "    \n",
    "    # display(eflux2_model.summary())\n",
    "    # display(eflux2_model.summary(solution=EFlux2_solution))\n",
    "\n",
    "    print('E-Flux2 status', EFlux2_solution.status)\n",
    "    print('E-Flux2 solution', EFlux2_solution.objective_value)\n",
    "    print()\n",
    "        \n",
    "    return EFlux2_solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4scRNA control\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpgs6r7cpm.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5164341231334846\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4443.086473084718\n",
      "\n",
      "CRISPRa-PP_1075\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp259nn7a9.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5104954646532736\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4428.3802720676695\n",
      "\n",
      "CRISPRa-PP_1830\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp9fkalps8.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5277894876884197\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4471.283158120534\n",
      "\n",
      "CRISPRa-PP_2168\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp1qmgnpfd.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.506830198872518\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4419.317558582417\n",
      "\n",
      "CRISPRa-PP_2329\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpqfgudbhq.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.514392709280912\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4438.028104288361\n",
      "\n",
      "CRISPRa-PP_4169\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpdas6j786.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5108434329235982\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4429.241202535737\n",
      "\n",
      "CRISPRa-PP_4965\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp2kcyx16v.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5187075158329406\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4448.723493714192\n",
      "\n",
      "CRISPRa-PP_5079\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmphpt_xuh7.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5154366140918867\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4440.614362811086\n",
      "\n",
      "CRISPRi-PP_0059\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmptbpyiusq.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5730485607153468\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4584.669008052766\n",
      "\n",
      "CRISPRi-PP_0289\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpwnyqoc9p.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5664139656944815\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4567.94736388948\n",
      "\n",
      "CRISPRi-PP_0417\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp2seqvwke.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5354215917508791\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4490.291244043726\n",
      "\n",
      "CRISPRi-PP_0420\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp12jp41sy.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5103678762874875\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4428.064621297862\n",
      "\n",
      "CRISPRi-PP_0966\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpntb3c06h.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4956140216257126\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4391.649801380317\n",
      "\n",
      "CRISPRi-PP_1025\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp07p8wuax.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5374514066611074\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4495.354257860111\n",
      "\n",
      "CRISPRi-PP_1032\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp678dpco0.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4910324393997685\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4380.376393718973\n",
      "\n",
      "CRISPRi-PP_1037\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpl4p1fnhs.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4759654132947009\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4343.41838742202\n",
      "\n",
      "CRISPRi-PP_1237\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpv_oznjna.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.526513604030562\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4468.109960572499\n",
      "\n",
      "CRISPRi-PP_1525\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpy9gy6tp3.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5357811589635482\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4491.187885846488\n",
      "\n",
      "CRISPRi-PP_1530\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpuzretpbf.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5306892232744604\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4478.4997072304495\n",
      "\n",
      "CRISPRi-PP_1665\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpvgobshio.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4755826481973435\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4342.481813733004\n",
      "\n",
      "CRISPRi-PP_1914\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp7q8akp0t.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5184755369860572\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4448.148102277703\n",
      "\n",
      "CRISPRi-PP_1917\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmppgz49a01.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4920647452683988\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4382.915044711757\n",
      "\n",
      "CRISPRi-PP_1988\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpu_m5z96p.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.468542090194437\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4325.274978901341\n",
      "\n",
      "CRISPRi-PP_1989\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpkb6zs0kc.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4930158585406201\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4385.254764472122\n",
      "\n",
      "CRISPRi-PP_1992\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpneop01c3.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5490503490052703\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4524.347603030873\n",
      "\n",
      "CRISPRi-PP_2000\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpujkks887.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4999752239471178\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4402.396200447729\n",
      "\n",
      "CRISPRi-PP_2082\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp6zif18f6.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5274763162451275\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4470.504164419876\n",
      "\n",
      "CRISPRi-PP_2554\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpbgum0vg1.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.512757258410385\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4433.978013409059\n",
      "\n",
      "CRISPRi-PP_2639\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp4a7n74r_.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5000912133705595\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4402.68221207682\n",
      "\n",
      "CRISPRi-PP_3569\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp7e0fdfyu.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4740515878079143\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4338.736665210048\n",
      "\n",
      "CRISPRi-PP_4016\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp125utxpu.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4941293570056597\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4387.994847885388\n",
      "\n",
      "CRISPRi-PP_4297\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp5_5u7yuu.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5198674100673568\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4451.6010824300765\n",
      "\n",
      "CRISPRi-PP_4678\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpiyf7w_ih.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5089180084944671\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4424.478575404031\n",
      "\n",
      "CRISPRi-PP_4725\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpaznv8xx6.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.541835806867201\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4506.301368700082\n",
      "\n",
      "CRISPRi-PP_4822\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmprtkpux1d.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4925982966162303\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4384.227483279334\n",
      "\n",
      "CRISPRi-PP_4823\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp8r525x73.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4731352713627253\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4336.496127816317\n",
      "\n",
      "CRISPRi-PP_5015\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpe3j66x_c.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5310139936600973\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4479.308370385441\n",
      "\n",
      "CRISPRi-PP_5128\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpde1u6dtw.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.489350592759865\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4376.242174953435\n",
      "\n",
      "CRISPRi-PP_5155\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp8fs9ekgq.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.4736456248258685\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4337.743941014135\n",
      "\n",
      "CRISPRi-PP_5336-5335\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmplpqv8xlf.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.493305832099224\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4385.9682344466355\n",
      "\n",
      "pACA production 3 scRNA positive control\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp_z3in9f4.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FBA solution: 1.5334613704947155\n",
      "E-Flux2 status optimal\n",
      "E-Flux2 solution 4485.404878841484\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fluxes = {}\n",
    "for strain in norm_data.columns:\n",
    "    print(strain)\n",
    "    fluxes[strain] = runEflux2(norm_data[strain], model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "efluxes = pd.DataFrame(dict([(rep, fluxes[rep].to_frame()['fluxes']) for rep in fluxes]))\n",
    "efluxes.to_csv('../data/iJN1463_eflux2_flux.csv')\n",
    "\n",
    "reduced_costs = pd.DataFrame(dict([(rep, fluxes[rep].to_frame()['reduced_costs']) for rep in fluxes]))\n",
    "# reduced_costs.to_csv('core66_eflux2_yield_reduced_costs.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(637, 41)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>4scRNA control</th>\n",
       "      <th>CRISPRa-PP_1075</th>\n",
       "      <th>CRISPRa-PP_1830</th>\n",
       "      <th>CRISPRa-PP_2168</th>\n",
       "      <th>CRISPRa-PP_2329</th>\n",
       "      <th>CRISPRa-PP_4169</th>\n",
       "      <th>CRISPRa-PP_4965</th>\n",
       "      <th>CRISPRa-PP_5079</th>\n",
       "      <th>CRISPRi-PP_0059</th>\n",
       "      <th>CRISPRi-PP_0289</th>\n",
       "      <th>...</th>\n",
       "      <th>CRISPRi-PP_4297</th>\n",
       "      <th>CRISPRi-PP_4678</th>\n",
       "      <th>CRISPRi-PP_4725</th>\n",
       "      <th>CRISPRi-PP_4822</th>\n",
       "      <th>CRISPRi-PP_4823</th>\n",
       "      <th>CRISPRi-PP_5015</th>\n",
       "      <th>CRISPRi-PP_5128</th>\n",
       "      <th>CRISPRi-PP_5155</th>\n",
       "      <th>CRISPRi-PP_5336-5335</th>\n",
       "      <th>pACA production 3 scRNA positive control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3HAD160</th>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "      <td>0.010419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1PQQS</th>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2PQQS</th>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3HAACOAT140</th>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "      <td>-0.010667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3HAACOAT141</th>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.010667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cinm_e</th>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_tyrA</th>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>...</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "      <td>0.039751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAL_Y</th>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_pHCA_Tr</th>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_pHCA_e</th>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>637 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             4scRNA control  CRISPRa-PP_1075  CRISPRa-PP_1830  \\\n",
       "3HAD160            0.010419         0.010419         0.010419   \n",
       "1PQQS              0.000065         0.000065         0.000065   \n",
       "2PQQS              0.000065         0.000065         0.000065   \n",
       "3HAACOAT140       -0.010667        -0.010667        -0.010667   \n",
       "3HAACOAT141        0.010667         0.010667         0.010667   \n",
       "...                     ...              ...              ...   \n",
       "EX_cinm_e          0.000010         0.000010         0.000010   \n",
       "R_tyrA             0.039751         0.039751         0.039751   \n",
       "PAL_Y              0.000010         0.000010         0.000010   \n",
       "R_pHCA_Tr          0.000010         0.000010         0.000010   \n",
       "EX_pHCA_e          0.000010         0.000010         0.000010   \n",
       "\n",
       "             CRISPRa-PP_2168  CRISPRa-PP_2329  CRISPRa-PP_4169  \\\n",
       "3HAD160             0.010419         0.010419         0.010419   \n",
       "1PQQS               0.000065         0.000065         0.000065   \n",
       "2PQQS               0.000065         0.000065         0.000065   \n",
       "3HAACOAT140        -0.010667        -0.010667        -0.010667   \n",
       "3HAACOAT141         0.010667         0.010667         0.010667   \n",
       "...                      ...              ...              ...   \n",
       "EX_cinm_e           0.000010         0.000010         0.000010   \n",
       "R_tyrA              0.039751         0.039751         0.039751   \n",
       "PAL_Y               0.000010         0.000010         0.000010   \n",
       "R_pHCA_Tr           0.000010         0.000010         0.000010   \n",
       "EX_pHCA_e           0.000010         0.000010         0.000010   \n",
       "\n",
       "             CRISPRa-PP_4965  CRISPRa-PP_5079  CRISPRi-PP_0059  \\\n",
       "3HAD160             0.010419         0.010419         0.010419   \n",
       "1PQQS               0.000065         0.000065         0.000065   \n",
       "2PQQS               0.000065         0.000065         0.000065   \n",
       "3HAACOAT140        -0.010667        -0.010667        -0.010667   \n",
       "3HAACOAT141         0.010667         0.010667         0.010667   \n",
       "...                      ...              ...              ...   \n",
       "EX_cinm_e           0.000010         0.000010         0.000010   \n",
       "R_tyrA              0.039751         0.039751         0.039751   \n",
       "PAL_Y               0.000010         0.000010         0.000010   \n",
       "R_pHCA_Tr           0.000010         0.000010         0.000010   \n",
       "EX_pHCA_e           0.000010         0.000010         0.000010   \n",
       "\n",
       "             CRISPRi-PP_0289  ...  CRISPRi-PP_4297  CRISPRi-PP_4678  \\\n",
       "3HAD160             0.010419  ...         0.010419         0.010419   \n",
       "1PQQS               0.000065  ...         0.000065         0.000065   \n",
       "2PQQS               0.000065  ...         0.000065         0.000065   \n",
       "3HAACOAT140        -0.010667  ...        -0.010667        -0.010667   \n",
       "3HAACOAT141         0.010667  ...         0.010667         0.010667   \n",
       "...                      ...  ...              ...              ...   \n",
       "EX_cinm_e           0.000010  ...         0.000010         0.000010   \n",
       "R_tyrA              0.039751  ...         0.039751         0.039751   \n",
       "PAL_Y               0.000010  ...         0.000010         0.000010   \n",
       "R_pHCA_Tr           0.000010  ...         0.000010         0.000010   \n",
       "EX_pHCA_e           0.000010  ...         0.000010         0.000010   \n",
       "\n",
       "             CRISPRi-PP_4725  CRISPRi-PP_4822  CRISPRi-PP_4823  \\\n",
       "3HAD160             0.010419         0.010419         0.010419   \n",
       "1PQQS               0.000065         0.000065         0.000065   \n",
       "2PQQS               0.000065         0.000065         0.000065   \n",
       "3HAACOAT140        -0.010667        -0.010667        -0.010667   \n",
       "3HAACOAT141         0.010667         0.010667         0.010667   \n",
       "...                      ...              ...              ...   \n",
       "EX_cinm_e           0.000010         0.000010         0.000010   \n",
       "R_tyrA              0.039751         0.039751         0.039751   \n",
       "PAL_Y               0.000010         0.000010         0.000010   \n",
       "R_pHCA_Tr           0.000010         0.000010         0.000010   \n",
       "EX_pHCA_e           0.000010         0.000010         0.000010   \n",
       "\n",
       "             CRISPRi-PP_5015  CRISPRi-PP_5128  CRISPRi-PP_5155  \\\n",
       "3HAD160             0.010419         0.010419         0.010419   \n",
       "1PQQS               0.000065         0.000065         0.000065   \n",
       "2PQQS               0.000065         0.000065         0.000065   \n",
       "3HAACOAT140        -0.010667        -0.010667        -0.010667   \n",
       "3HAACOAT141         0.010667         0.010667         0.010667   \n",
       "...                      ...              ...              ...   \n",
       "EX_cinm_e           0.000010         0.000010         0.000010   \n",
       "R_tyrA              0.039751         0.039751         0.039751   \n",
       "PAL_Y               0.000010         0.000010         0.000010   \n",
       "R_pHCA_Tr           0.000010         0.000010         0.000010   \n",
       "EX_pHCA_e           0.000010         0.000010         0.000010   \n",
       "\n",
       "             CRISPRi-PP_5336-5335  pACA production 3 scRNA positive control  \n",
       "3HAD160                  0.010419                                  0.010419  \n",
       "1PQQS                    0.000065                                  0.000065  \n",
       "2PQQS                    0.000065                                  0.000065  \n",
       "3HAACOAT140             -0.010667                                 -0.010667  \n",
       "3HAACOAT141              0.010667                                  0.010667  \n",
       "...                           ...                                       ...  \n",
       "EX_cinm_e                0.000010                                  0.000010  \n",
       "R_tyrA                   0.039751                                  0.039751  \n",
       "PAL_Y                    0.000010                                  0.000010  \n",
       "R_pHCA_Tr                0.000010                                  0.000010  \n",
       "EX_pHCA_e                0.000010                                  0.000010  \n",
       "\n",
       "[637 rows x 41 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(efluxes.shape)\n",
    "efluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing Eflux2 results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "efluxResults = efluxes.round(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are the reactions whose flux solutions vary by strain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "varEflux2 = efluxResults[efluxResults.var(axis=1)>1E-5].index\n",
    "\n",
    "noEX_varEflux2 = [i for i in varEflux2 if 'EX_' not in i]\n",
    "\n",
    "# check if reaction has a gene reaction rule\n",
    "noEX_varEflux2 = [i for i in noEX_varEflux2 if model.reactions.get_by_id(i).gene_reaction_rule]\n",
    "# remove RgPAL\n",
    "# noEX_varEflux2.remove('RgPAL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the transcriptomics values for reactions listed in noEX_varEflux2\n",
    "ds = []\n",
    "for strain in norm_data.columns:\n",
    "    # get transcriptionFactors for each reaction\n",
    "    transcriptFactors = [transcript_value_for_rxn(model, norm_data[strain], model.reactions.get_by_id(i)) for i in noEX_varEflux2]\n",
    "    transcriptFactors = dict(zip(noEX_varEflux2, transcriptFactors))\n",
    "    # get rid of reactions with infinite transcription\n",
    "    transcriptFactors = {k: v for k, v in transcriptFactors.items() if v != np.inf}\n",
    "    ds.append(transcriptFactors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strains</th>\n",
       "      <th>4scRNA control</th>\n",
       "      <th>CRISPRa-PP_1075</th>\n",
       "      <th>CRISPRa-PP_1830</th>\n",
       "      <th>CRISPRa-PP_2168</th>\n",
       "      <th>CRISPRa-PP_2329</th>\n",
       "      <th>CRISPRa-PP_4169</th>\n",
       "      <th>CRISPRa-PP_4965</th>\n",
       "      <th>CRISPRa-PP_5079</th>\n",
       "      <th>CRISPRi-PP_0059</th>\n",
       "      <th>CRISPRi-PP_0289</th>\n",
       "      <th>...</th>\n",
       "      <th>CRISPRi-PP_4297</th>\n",
       "      <th>CRISPRi-PP_4678</th>\n",
       "      <th>CRISPRi-PP_4725</th>\n",
       "      <th>CRISPRi-PP_4822</th>\n",
       "      <th>CRISPRi-PP_4823</th>\n",
       "      <th>CRISPRi-PP_5015</th>\n",
       "      <th>CRISPRi-PP_5128</th>\n",
       "      <th>CRISPRi-PP_5155</th>\n",
       "      <th>CRISPRi-PP_5336-5335</th>\n",
       "      <th>pACA production 3 scRNA positive control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDPA</th>\n",
       "      <td>0.889054</td>\n",
       "      <td>0.914232</td>\n",
       "      <td>0.948481</td>\n",
       "      <td>0.925574</td>\n",
       "      <td>0.857283</td>\n",
       "      <td>0.929645</td>\n",
       "      <td>0.948159</td>\n",
       "      <td>0.927870</td>\n",
       "      <td>0.912316</td>\n",
       "      <td>0.926168</td>\n",
       "      <td>...</td>\n",
       "      <td>0.908321</td>\n",
       "      <td>0.907644</td>\n",
       "      <td>0.879046</td>\n",
       "      <td>0.807352</td>\n",
       "      <td>0.807937</td>\n",
       "      <td>0.873997</td>\n",
       "      <td>0.902806</td>\n",
       "      <td>0.804419</td>\n",
       "      <td>0.808418</td>\n",
       "      <td>0.920279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_PABASYN_RXN</th>\n",
       "      <td>0.980204</td>\n",
       "      <td>0.957497</td>\n",
       "      <td>0.995524</td>\n",
       "      <td>0.974859</td>\n",
       "      <td>0.933387</td>\n",
       "      <td>0.968044</td>\n",
       "      <td>0.966240</td>\n",
       "      <td>0.979097</td>\n",
       "      <td>0.953145</td>\n",
       "      <td>0.982304</td>\n",
       "      <td>...</td>\n",
       "      <td>0.970946</td>\n",
       "      <td>0.962346</td>\n",
       "      <td>0.967185</td>\n",
       "      <td>0.912694</td>\n",
       "      <td>0.913534</td>\n",
       "      <td>0.945290</td>\n",
       "      <td>0.958328</td>\n",
       "      <td>0.901699</td>\n",
       "      <td>0.915672</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN_20192</th>\n",
       "      <td>0.890601</td>\n",
       "      <td>0.949730</td>\n",
       "      <td>0.978755</td>\n",
       "      <td>0.911726</td>\n",
       "      <td>0.988292</td>\n",
       "      <td>0.911649</td>\n",
       "      <td>0.932567</td>\n",
       "      <td>0.927089</td>\n",
       "      <td>0.948705</td>\n",
       "      <td>0.913962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.928289</td>\n",
       "      <td>0.932839</td>\n",
       "      <td>0.988968</td>\n",
       "      <td>0.980697</td>\n",
       "      <td>0.992602</td>\n",
       "      <td>0.894180</td>\n",
       "      <td>0.912959</td>\n",
       "      <td>0.952851</td>\n",
       "      <td>0.977599</td>\n",
       "      <td>0.973714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RXN_20193</th>\n",
       "      <td>0.985857</td>\n",
       "      <td>0.949793</td>\n",
       "      <td>0.978540</td>\n",
       "      <td>0.929860</td>\n",
       "      <td>0.950992</td>\n",
       "      <td>0.948174</td>\n",
       "      <td>0.936890</td>\n",
       "      <td>0.956916</td>\n",
       "      <td>0.975732</td>\n",
       "      <td>0.976080</td>\n",
       "      <td>...</td>\n",
       "      <td>0.957520</td>\n",
       "      <td>0.951945</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931961</td>\n",
       "      <td>0.926160</td>\n",
       "      <td>0.943306</td>\n",
       "      <td>0.943818</td>\n",
       "      <td>0.918894</td>\n",
       "      <td>0.927636</td>\n",
       "      <td>0.974297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RgPAL</th>\n",
       "      <td>0.964010</td>\n",
       "      <td>0.960234</td>\n",
       "      <td>0.971228</td>\n",
       "      <td>0.957904</td>\n",
       "      <td>0.962712</td>\n",
       "      <td>0.960456</td>\n",
       "      <td>0.965455</td>\n",
       "      <td>0.963376</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.966192</td>\n",
       "      <td>0.959232</td>\n",
       "      <td>0.980158</td>\n",
       "      <td>0.948857</td>\n",
       "      <td>0.936484</td>\n",
       "      <td>0.973278</td>\n",
       "      <td>0.946793</td>\n",
       "      <td>0.936809</td>\n",
       "      <td>0.949307</td>\n",
       "      <td>0.974834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "strains        4scRNA control  CRISPRa-PP_1075  CRISPRa-PP_1830  \\\n",
       "DDPA                 0.889054         0.914232         0.948481   \n",
       "R_PABASYN_RXN        0.980204         0.957497         0.995524   \n",
       "RXN_20192            0.890601         0.949730         0.978755   \n",
       "RXN_20193            0.985857         0.949793         0.978540   \n",
       "RgPAL                0.964010         0.960234         0.971228   \n",
       "\n",
       "strains        CRISPRa-PP_2168  CRISPRa-PP_2329  CRISPRa-PP_4169  \\\n",
       "DDPA                  0.925574         0.857283         0.929645   \n",
       "R_PABASYN_RXN         0.974859         0.933387         0.968044   \n",
       "RXN_20192             0.911726         0.988292         0.911649   \n",
       "RXN_20193             0.929860         0.950992         0.948174   \n",
       "RgPAL                 0.957904         0.962712         0.960456   \n",
       "\n",
       "strains        CRISPRa-PP_4965  CRISPRa-PP_5079  CRISPRi-PP_0059  \\\n",
       "DDPA                  0.948159         0.927870         0.912316   \n",
       "R_PABASYN_RXN         0.966240         0.979097         0.953145   \n",
       "RXN_20192             0.932567         0.927089         0.948705   \n",
       "RXN_20193             0.936890         0.956916         0.975732   \n",
       "RgPAL                 0.965455         0.963376         1.000000   \n",
       "\n",
       "strains        CRISPRi-PP_0289  ...  CRISPRi-PP_4297  CRISPRi-PP_4678  \\\n",
       "DDPA                  0.926168  ...         0.908321         0.907644   \n",
       "R_PABASYN_RXN         0.982304  ...         0.970946         0.962346   \n",
       "RXN_20192             0.913962  ...         0.928289         0.932839   \n",
       "RXN_20193             0.976080  ...         0.957520         0.951945   \n",
       "RgPAL                 0.995782  ...         0.966192         0.959232   \n",
       "\n",
       "strains        CRISPRi-PP_4725  CRISPRi-PP_4822  CRISPRi-PP_4823  \\\n",
       "DDPA                  0.879046         0.807352         0.807937   \n",
       "R_PABASYN_RXN         0.967185         0.912694         0.913534   \n",
       "RXN_20192             0.988968         0.980697         0.992602   \n",
       "RXN_20193             1.000000         0.931961         0.926160   \n",
       "RgPAL                 0.980158         0.948857         0.936484   \n",
       "\n",
       "strains        CRISPRi-PP_5015  CRISPRi-PP_5128  CRISPRi-PP_5155  \\\n",
       "DDPA                  0.873997         0.902806         0.804419   \n",
       "R_PABASYN_RXN         0.945290         0.958328         0.901699   \n",
       "RXN_20192             0.894180         0.912959         0.952851   \n",
       "RXN_20193             0.943306         0.943818         0.918894   \n",
       "RgPAL                 0.973278         0.946793         0.936809   \n",
       "\n",
       "strains        CRISPRi-PP_5336-5335  pACA production 3 scRNA positive control  \n",
       "DDPA                       0.808418                                  0.920279  \n",
       "R_PABASYN_RXN              0.915672                                  1.000000  \n",
       "RXN_20192                  0.977599                                  0.973714  \n",
       "RXN_20193                  0.927636                                  0.974297  \n",
       "RgPAL                      0.949307                                  0.974834  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boundFactors = pd.DataFrame(ds, index=norm_data.columns).transpose()\n",
    "boundFactors.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpxc6dk5tr.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ACONTa\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp4ynizvws.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ACONTb\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp7krtr6zx.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "CS\n",
      "FBA solution: 1.4679376520626506\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmphco4w2xg.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "CYTBO3_4pp\n",
      "FBA solution: None\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\user\\miniconda3\\envs\\new_Gayles\\lib\\site-packages\\cobra\\util\\solver.py:554: UserWarning: Solver status is 'infeasible'.\n",
      "  warn(f\"Solver status is '{status}'.\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp1mr25csx.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "DHQS\n",
      "FBA solution: 1.3418506876585041\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpxy1slevi.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "EDA\n",
      "FBA solution: 0.5699911577423645\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpncdc1neu.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "EDD\n",
      "FBA solution: 0.8066169417338137\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp35yj6exh.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ENO\n",
      "FBA solution: 1.4983990609196456\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp9sbf5mpm.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "FUM\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpsvfc433r.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "G3PD2\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp39d04mfm.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLCDpp\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpfee8lkai.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLCNt2rpp\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpgloup4v3.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLCtex\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpxybzjgy7.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLNS_copy1\n",
      "FBA solution: 1.4603596699246473\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpjr2m3s_1.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLUDxi\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpst39us1f.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLUDy\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpi5ab8zw7.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLXCL\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp6yd4rib9.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLYALDDr\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpraw1gl4x.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GLYK\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpa335tyck.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GND\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpeot568ga.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GNK\n",
      "FBA solution: 1.246418951134747\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpgveew9q0.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "HPYRI\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpcdi6l_xi.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "HPYRRy\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpktk12mxr.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ICDHyr\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpknk9a3lw.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ICL\n",
      "FBA solution: 1.474802775972429\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpcuf0y4cz.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "MDH\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpjo9kh5pc.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "NADH16pp\n",
      "FBA solution: None\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\user\\miniconda3\\envs\\new_Gayles\\lib\\site-packages\\cobra\\util\\solver.py:554: UserWarning: Solver status is 'infeasible'.\n",
      "  warn(f\"Solver status is '{status}'.\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp_zjv0iw6.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "NH4tpp\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpjylvglsd.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "PDH\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpup_p02z3.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "PGK\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp6iv92afi.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "PGM\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpkkgcf8z3.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "RPI\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpz5d6yuef.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "SHK3Dr\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpowmja8_s.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "SHKK\n",
      "FBA solution: 1.3654668809660522\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpv7jwi5ct.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "SUCDi\n",
      "FBA solution: 1.4603738234600279\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpkfidztj2.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "SUCOAS\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp0bv_5iww.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "TALA\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp2x6vsbq7.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "TKT1\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmphdsqbcer.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "TKT2\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpvvxegfdh.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "TPI\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmph0wznz3y.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "DHQTi_copy2\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpeufdo2nv.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "PSCVT_copy2\n",
      "FBA solution: 1.3095680899084787\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpe2p6o270.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "GAPDi_nadp\n",
      "FBA solution: None\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\user\\miniconda3\\envs\\new_Gayles\\lib\\site-packages\\cobra\\util\\solver.py:554: UserWarning: Solver status is 'infeasible'.\n",
      "  warn(f\"Solver status is '{status}'.\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp_06fekpk.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "AKGDH\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpferw6b5d.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ALCD19\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpru86g0yh.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "ATPS4rpp\n",
      "FBA solution: None\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\user\\miniconda3\\envs\\new_Gayles\\lib\\site-packages\\cobra\\util\\solver.py:554: UserWarning: Solver status is 'infeasible'.\n",
      "  warn(f\"Solver status is '{status}'.\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmplhjqq7ri.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "CHORS\n",
      "FBA solution: 1.4151752108028648\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpiiol6qm9.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "DDPA\n",
      "FBA solution: 1.385375629423771\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmp0va_o93h.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "R_PABASYN_RXN\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpkiyf7lcl.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "RXN_20192\n",
      "FBA solution: 1.400957196883513\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpl48xk8x6.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "RXN_20193\n",
      "FBA solution: 1.5164345430151993\n",
      "Read LP format model from file C:\\Users\\user\\AppData\\Local\\Temp\\tmpp78ecxm7.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 671 rows, 1274 columns, 5766 nonzeros\n",
      "RgPAL\n",
      "FBA solution: 1.5164341231334866\n"
     ]
    }
   ],
   "source": [
    "# for each reaction in boundFactors['4scRNA control'], \n",
    "for rxn in boundFactors.index:\n",
    "    model3 = model.copy()\n",
    "    sol = model3.optimize()\n",
    "    model3.reactions.RgPAL.upper_bound = sol['RgPAL'] * 0.964010\n",
    "    print(rxn)\n",
    "    # multiply the upperbound by the bound Factor\n",
    "    model3.reactions.get_by_id(rxn).upper_bound = sol[rxn] * boundFactors['4scRNA control'][rxn]\n",
    "    # run FBA\n",
    "    model3sol = model3.optimize()\n",
    "    print('FBA solution:', model3sol.objective_value)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modeling-workshop",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": false,
   "autoclose": false,
   "autocomplete": false,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "ctrl-q"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": true,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
